cmd/checkpoint/main.go

```diff
@@ -0,0 +1,157 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"k8s.io/kubernetes/pkg/api/v1"
+)
+
+var manifestBytes = []byte(`
+{
+"apiVersion": "v1",
+"kind": "Pod",
+"metadata": {
+  "name": "temp-apiserver",
+  "namespace": "kube-system"
+},
+"spec": {}
+}
+`)
+
+var (
+	tempAPIServer = []byte("temp-apiserver")
+	kubeAPIServer = []byte("kube-apiserver")
+)
+
+const (
+	kubeletAPIPodsURL = "http://localhost:10255/pods"
+	ignorePath        = "/etc/kubernetes/_manifests/apiserver.json"
+	activePath        = "/etc/kubernetes/manifests/apiserver.json"
+)
+
+func main() {
+	log.Println("begin apiserver checkpointing...")
+	// Create ignore dir.
+	if err := os.MkdirAll(filepath.Dir(ignorePath), 0755); err != nil {
+		log.Fatal(err)
+	}
+	run()
+}
+
+func run() {
+	var manifest v1.Pod
+	if err := json.Unmarshal(manifestBytes, &manifest); err != nil {
+		log.Fatal(err)
+	}
+	for {
+		rawPods := getPods()
+		switch {
+		case bothAPIServersRunning(rawPods):
+			log.Println("both temp and kube apiserver running, removing temp apiserver")
+			// Both the self-hosted API Server and the temp API Server are running.
+			// Remove the temp API Server manifest from the config dir so that the
+			// kubelet will stop it.
+			if err := os.Remove(activePath); err != nil {
+				log.Println(err)
+			}
+		case kubeSystemAPIServerRunning(rawPods):
+			log.Println("kube-apiserver found, creating temp-apiserver manifest")
+			// The self-hosted API Server is running. Let's snapshot the pod,
+			// clean it up a bit, and then save it to the ignore path for
+			// later use.
+			var podList v1.PodList
+			var apiPod v1.Pod
+			if err := json.Unmarshal(rawPods, &podList); err != nil {
+				log.Fatal(err)
+			}
+			for _, p := range podList.Items {
+				if strings.Contains(p.Name, string(kubeAPIServer)) {
+					apiPod = p
+					break
+				}
+			}
+			cleanVolumes(&apiPod)
+			modifyInsecurePort(&apiPod)
+			manifest.Spec = apiPod.Spec
+			m, err := json.Marshal(manifest)
+			if err != nil {
+				log.Fatal(err)
+			}
+			if err := ioutil.WriteFile(ignorePath, m, 0644); err != nil {
+				log.Fatal(err)
+			}
+
+		default:
+			log.Println("no apiserver running, installing temp apiserver static manifest")
+			b, err := ioutil.ReadFile(ignorePath)
+			if err != nil {
+				log.Println(err)
+				continue
+			}
+			if err := ioutil.WriteFile(activePath, b, 0644); err != nil {
+				log.Println(err)
+			}
+		}
+		time.Sleep(time.Second)
```

Comment:
Sure. Ideally we could do something like watch the kubelet-api (like the apis-server) and only act on pod changing -- but I'm not sure that the kubelet-api would provide something like that.

Alternatively, we could do something like hash the spec and if it's changed then re-checkpoint rather than blindly doing it every second. But I'm okay with this being a follow-up optimization.

But I do think we should give the api-server some time to recover on it's own. There already are some concepts of checkpointing via docker labels (not sure about rkt) so that if the container dies, kubelet is responsible for starting it again (e.g. api-server). We just want to fail back to our checkpoint if that doesn't occur.